Fix findfirst syntax #89

kescobo opened the pull request:
Trying to address #89 

I think I did what @oxinabox [mentioned here](https://github.com/oxinabox/DataDeps.jl/issues/89#issuecomment-477492727), but my attempt at testing doesn't seem to throw the error. 

Also, there are a bunch of other tests failing on `1.1` even on master, and it looks like travis is only testing on 0.7 and nightly, so I can't tell if my change screwed anything else up. 

cc @Ayushk4

oxinabox commented on src/locations.jl:
```diff
@@ -109,10 +109,8 @@ Determines the location to save a datadep with the given name to.
-        0 == first(uv_access(path, W_OK))
-    end
-    if path_ind==0
+    path_ind = findfirst(can_read_file, cands)
```
This is not the same
`can_read_file` checks for `R_OK`,
but we need to check `W_OK`, 
like in `uv_access(path, W_OK)`

kescobo replied:
Got it - reverted

oxinabox commented on test/locations.jl:
```diff
@@ -10,5 +10,11 @@ using DataDeps: try_determine_load_path, determine_save_path, try_determine_pack
-end
+    mktemp() do fn, fh
+        ENV["DATADEPS_LOAD_PATH"] = ""
+        ENV["DATADEPS_NO_STANDARD_LOAD_PATH"] = false
```
`DATADEPS_NO_STANDARD_LOAD_PATH`
should be set to "true".
As this disables the standard locations

(sorry I said it wrong in the issue).

Also use `withenv(...) do` so that the enviroment variable setting is scoped.
Or it will break all the other tests, since they will not have anywhere to save.

kescobo replied:
Check check

oxinabox commented on test/locations.jl:
```diff
@@ -10,5 +10,11 @@ using DataDeps: try_determine_load_path, determine_save_path, try_determine_pack
+    mktemp() do fn, fh
+        ENV["DATADEPS_LOAD_PATH"] = ""
+        ENV["DATADEPS_NO_STANDARD_LOAD_PATH"] = false
+        @test try_determine_package_datadeps_dir(fn) == nothing
```
Why is this test being added?
I'm not sure I see the connection to #89 

At a guess maybe it was mistakenly duplicated from above?

kescobo replied:
Yeah, I was just checking that my method of testing locally was actually testing the right thing by adding an extra test that I knew worked, Removed now

oxinabox commented on test/locations.jl:
```diff
@@ -10,5 +10,11 @@ using DataDeps: try_determine_load_path, determine_save_path, try_determine_pack
     end
 
-end
+    mktemp() do fn, fh
```
I do not think this `mktemp()` is required

kescobo replied:
You're right, again duplicated from above. Removed

oxinabox commented:
If you rebase against master, I have fixed the failing test and enabled more CI

kescobo commented:
OK, I've rebased on master and I think made the fixes you mentioned. However, the `@test_throws` I added passes even with the code from `master`, so I don't think it's doing the right thing. Do I need to pass a particular path to `determine_save_path()`?

oxinabox commented on Manifest.toml:
```diff
@@ -0,0 +1,95 @@
+# This file is machine-generated - editing it directly is not advised
```
Seem to have version controlled the Manifest.toml by mistake?

oxinabox commented on test/locations.jl:
```diff
@@ -10,5 +10,8 @@ using DataDeps: try_determine_load_path, determine_save_path, try_determine_pack
 
-end
+    withenv("DATADEPS_LOAD_PATH" => "", "DATADEPS_NO_STANDARD_LOAD_PATH" => true) do
+        @test_throws NoValidPathError determine_save_path("test")
```
```suggestion
        @test_throws NoValidPathError determine_save_path("DummyDataDepName")
```

oxinabox commented:
> the @test_throws I added passes even with the code from master, so I don't think it's doing the right thing.

Hmmm, it looks right to me. Will probably require some debugging and local testing. Either by adding some print statements, or by using the fancy new debugger. 

> Do I need to pass a particular path to determine_save_path()?

Nope, just the name of the DataDep itself.


> And thanks for taking the time to review and walk me through it - I know it's probably more time than it would have taken to just fix it yourself :-)

No problem, it is an investment. :-)

kescobo commented:
After some investigation (Debugger is pretty cool!), it turns out I'm full of crap. Test fails on `master` but passes on this branch, so I think :shipit:

I must need some work with my Revise Workflow. Thanks again for the shepherding!

oxinabox commented on test/locations.jl:
```diff
@@ -10,5 +10,8 @@ using DataDeps: try_determine_load_path, determine_save_path, try_determine_pack
     end
 
-end
+    withenv("DATADEPS_LOAD_PATH" => "", "DATADEPS_NO_STANDARD_LOAD_PATH" => true) do
```
```suggestion
    # Make sure we get the right error when no writeable path
    #  see #89
    withenv("DATADEPS_LOAD_PATH" => "", "DATADEPS_NO_STANDARD_LOAD_PATH" => "true") do
```

oxinabox commented:
Cool, some finishing touches:

Can you take a look at the 2 suggestions i made,
and also perhaps think if there is a better error message?

Something like _"No writable path exists to save the data, Make sure there exists as writable path in your DataDeps Load Path"_ and probably print a list of all the paths it checks?

kescobo commented:
How about out this?

oxinabox commented on src/locations.jl:
```diff
@@ -113,6 +113,7 @@ function determine_save_path(name, rel=nothing)::String
         0 == first(uv_access(path, W_OK))
     end
     if path_ind === nothing
+        @error "No writable path exists to save the data, make sure there exists as writable path in your DataDeps Load Path" ENV["DATADEPS_LOAD_PATH"]
```
`ENV["DATADEPS_LOAD_PATH"]` does not contain the actual complete loadpath.
`cands` does though, as that is what `preferred_paths` returns.
It should probably say "The current load path contains:"
since it can be changed

Also I am not sure about the capitalisation.
Nor one the pros and cons of using `@error` vs including this in the error message.

kescobo replied:
When I have error messages that are more than a couple of words (esp if I'm including calculated values), I just prefer the formatting of `@error`. Happy to put it all directly in the exception if you prefer.

oxinabox replied:
ok, lets try it your way.
Can change it later  if someone manages to get  the  error  but  not  the log  message

kescobo commented:
Printing for `NoValidPathError` isn't great, compare

```julia
julia> withenv("DATADEPS_LOAD_PATH" => "", "DATADEPS_NO_STANDARD_LOAD_PATH" => "true") do
               determine_save_path("DummyDataDepName")
           end
ERROR: NoValidPathError("No writable path exists to save the data. Make sure there exists as writable path in your DataDeps Load Path.\nSee http://white.ucc.asn.au/DataDeps.jl/stable/z10-for-end-users.html#The-Load-Path-1\nThe current load path contains: [\"\"]\n")
```

to this, if I change that to an `ErrorException`:

```julia
julia> withenv("DATADEPS_LOAD_PATH" => "", "DATADEPS_NO_STANDARD_LOAD_PATH" => "true") do
               determine_save_path("DummyDataDepName")
           end
ERROR: No writable path exists to save the data. Make sure there exists as writable path in your DataDeps Load Path.
See http://white.ucc.asn.au/DataDeps.jl/stable/z10-for-end-users.html#The-Load-Path-1
The current load path contains: [""]
```

oxinabox commented:
I'm  sold

kescobo commented:
I'm a bit confused by the codecov outputs, but I think this is gtg?

oxinabox commented:
CodeCov  seeems broken.
Tests pass so good to go
